Prototype for PyTorch-specific flake8 plugin

kit1980 opened the pull request:
The big vision for the plugin is to flag (and for some auto fix) all kinds of problematic usages of PyTorch.

What it is doing so far is printing error messages for the functions from the deprecated_symbols.yaml.
Eventually this list will be in the pytorch/pytorch and also will be used for generating documentation about deprecated and removed functions.

The plugin is in a working state, there is an output from checking torchaudio:

```
 flake8 --isolated --select=TOR
./examples/libtorchaudio/augmentation/create_jittable_pipeline.py:38:22: TOR101 Use of deprecated function torch.norm
./examples/tutorials/audio_data_augmentation_tutorial.py:193:14: TOR101 Use of deprecated function torch.norm
./test/torchaudio_unittest/transforms/transforms_test_impl.py:475:16: TOR101 Use of deprecated function torch.norm
./test/torchaudio_unittest/transforms/transforms_test_impl.py:477:30: TOR101 Use of deprecated function torch.norm
./test/torchaudio_unittest/transforms/transforms_test_impl.py:479:30: TOR101 Use of deprecated function torch.norm
./torchaudio/functional/functional.py:1039:27: TOR101 Use of deprecated function torch.norm
./torchaudio/functional/functional.py:1040:27: TOR101 Use of deprecated function torch.norm
```

ZainRizvi commented on tools/flake8-torch/flake8_torch/checker.py:
```diff
@@ -0,0 +1,96 @@
+        self.deprecated_config = {}
+        with open(Path(__file__).absolute().parent / "deprecated_symbols.yaml") as f:
+            for item in yaml.load(f, yaml.SafeLoader):
+                self.deprecated_config[item["name"]] = item
```
this could be a good place to check for duplicate items with the same name

kit1980 replied:
Maybe later.

ZainRizvi commented on tools/flake8-torch/flake8_torch/checker.py:
```diff
@@ -0,0 +1,96 @@
+    name = "flake8-torch"
+    version = "0.0.1"
+
+    def __init__(self, tree):
```
What is this a tree of, and where does it come from?  I bit of context on how linter plugins work would be helpful here

kit1980 replied:
This is an AST tree, should be self-explanatory in this context.

ZainRizvi commented on tools/flake8-torch/flake8_torch/checker.py:
```diff
@@ -0,0 +1,96 @@
+        for (name, lineno, col_offset) in self.func_calls:
+            if name in self.deprecated_config:
+                if self.deprecated_config[name]["remove_pr"] is None:
+                    message = f"TOR101 Use of deprecated function {name}"
```
I'd love to hear your thoughts on what the rollout process would look like when someone wants to deprecate a new function once this feature is in.  I'm especially curious about when build breaks are expected to happen, when (potentially ignorable) warnings would be given, and when issues/tasks would be filed.

An idealized form might look something like:
1. Dev A marks function torch.Foo as deprecated. As part of his PR, he runs a script to grandfather in all existing usages of that deprecated function, blocking new invocations.
2. Once the PR is merged, issues/tasks get filed against the code owners to remove those now-deprecated usages. (Perhaps in the future we could even offer tooling that auto-replace the usages)
3. As the deprecated function's invocations are replaced, keep shrinking the list of places that are allowed to use that deprecated function (clean up the list)
4. Until the list of invocations has shrunk to zero, any PR that attempts to mark the function as removed will fail

ZainRizvi reviewed (commented):
This is looking pretty promising!

I'd love to hear your thoughts on what the rollout process would look like when someone wants to deprecate a new function once this feature is in. 

I'm especially curious about when build breaks are expected to happen, when (potentially ignorable) warnings would be given, and when issues/tasks would be filed.

An idealized form might look something like:

1. Dev A marks function torch.Foo as deprecated. As part of his PR, he runs a script to grandfather in all existing usages of that deprecated function, blocking new invocations.
2. Once the PR is merged, issues/tasks get filed against the code owners to remove those now-deprecated usages. (Perhaps in the future we could even offer tooling that auto-replace the usages)
3. As the deprecated function's invocations are replaced, keep shrinking the list of places that are allowed to use that deprecated function (clean up the list)
4. Until the list of invocations has shrunk to zero, any PR that attempts to mark the function as removed will fail

kit1980 commented on tools/flake8-torch/flake8_torch/checker.py:
```diff
@@ -0,0 +1,96 @@
+                if self.deprecated_config[name]["remove_pr"] is None:
+                    message = f"TOR101 Use of deprecated function {name}"
+                else:
+                    message = f"TOR201 Use of removed function {name}"
```
I just came up with the codes, they are arbitrary.
The idea that the errors should be grouped logically in some way, so users can filter out classes of errors they are not interested in.

kit1980 commented on tools/flake8-torch/flake8_torch/deprecated_symbols.yaml:
```diff
@@ -0,0 +1,51 @@
+  deprecate_pr: https://github.com/pytorch/pytorch/pull/57745
+  remove_pr:
+
+- name: torch.symeig
+  deprecate_pr: https://github.com/pytorch/pytorch/pull/57732
+  remove_pr: https://github.com/pytorch/pytorch/pull/70988
```
This file will eventually live in pytorch/pytorch, and there will be probably another checker on pytorch/pytorch looking for PRs with module:deprecation label or even search for strings "deprecated".
We'll need to discuss this.
This prototype is mostly to facilitate the discussion, but it's also already useful as is to find issues.

malfet commented on tools/flake8-torch/tests/fixtures/deprecated_from_nn.py:
```diff
@@ -0,0 +1,8 @@
+from torch import nn as xx
+xx.UpsamplingNearest2d()
+
+import torch.nn as yy
+yy.UpsamplingNearest2d()
```
How about that?
```suggestion
import torch.nn as yy
yy.UpsamplingNearest2d()

import torch.nn.UpsamplingNearest2d as ZZZ
ZZZ()

FOO = ZZZ
FOO()
```

kit1980 replied:
`FOO = ZZZ; FOO()` will not be caught currently.
I don't think it's a very common pattern.
We can add that later if needed, may be a bit tricky to implement, but should be doable.

malfet commented on tools/flake8-torch/pyproject.toml:
```diff
@@ -0,0 +1,16 @@
+[project]
```
Please add dependency on `yaml`

malfet commented on tools/flake8-torch/flake8_torch/checker.py:
```diff
@@ -0,0 +1,96 @@
+
+class TorchChecker:
+    name = "flake8-torch"
+    version = "0.0.1"
```
Is it possible to keep versions in sync between pyproject.toml and this file?

kit1980 replied:
done

malfet commented on tools/flake8-torch/flake8_torch/checker.py:
```diff
@@ -0,0 +1,96 @@
+        module = "" if node.module is None else node.module
+        module = "." * node.level + module
+        for alias in node.names:
+            if alias.asname is None:
```
Is `asname` guaranteed to be defined? If not, use `getattr`
```suggestion
            if getattr(alias, 'asname', None) is None:
```

kit1980 replied:
`asname` is always defined, but can be None: https://docs.python.org/3/library/ast.html#ast.alias

malfet commented on tools/flake8-torch/flake8_torch/checker.py:
```diff
@@ -0,0 +1,96 @@
+        module = "." * node.level + module
+        for alias in node.names:
+            if alias.asname is None:
+                self._import_info[alias.name] = module + "." + alias.name
+            else:
+                self._import_info[alias.asname] = module + "." + alias.name
```
One f-string to rull them all...
```suggestion
                self._import_info[alias.name] = f"{module}.{alias.name}"
            else:
                self._import_info[alias.asname] = f"{module}.{alias.name}"
```

kit1980 replied:
Done

malfet reviewed (approved):
LGTM, but how about adding a CI workflow that'll test those?
Also, please consider addressing nits

malfet reviewed (approved):
LGTM, but how about adding a CI workflow that'll test those?
Also, please consider addressing nits

kit1980 commented on tools/flake8-torch/tests/test_flake8_torch.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def _fixture_path(filename):
+    return Path(__file__).absolute().parent / "fixtures" / filename
```
Removed that function, no longer needed.

kit1980 commented:
> LGTM, but how about adding a CI workflow that'll test those? Also, please consider addressing nits

I've updated the code to address the comments.
I'll add a workflow to run the tests in the next PR.
